x-pack/metricbeat/module/azure/monitor/client_helper.go

```diff
@@ -0,0 +1,137 @@
+func mapMetric(client *azure.Client, metric azure.MetricConfig, resource resources.GenericResource) ([]azure.Metric, error) {
+	var metrics []azure.Metric
+	// get all metrics supported by the namespace provided
+	metricDefinitions, err := client.AzureMonitorService.GetMetricDefinitions(*resource.ID, metric.Namespace)
+	if err != nil {
+		return nil, errors.Wrapf(err, "no metric definitions were found for resource %s and namespace %s.", *resource.ID, metric.Namespace)
+	}
+	if len(*metricDefinitions.Value) == 0 {
+		return nil, errors.Errorf("no metric definitions were found for resource %s and namespace %s.", *resource.ID, metric.Namespace)
+	}
+
+	// validate metric names
+	// check if all metric names are selected (*)
+	var supportedMetricNames []string
+	var unsupportedMetricNames []string
+	if azure.StringInSlice("*", metric.Name) {
+		for _, definition := range *metricDefinitions.Value {
+			supportedMetricNames = append(supportedMetricNames, *definition.Name.Value)
+		}
+	} else {
+		// verify if configured metric names are valid, return log error event for the invalid ones, map only  the valid metric names
+		supportedMetricNames, unsupportedMetricNames = filterMetrics(metric.Name, *metricDefinitions.Value)
+		if len(unsupportedMetricNames) > 0 {
+			return nil, errors.Errorf("none of metric names configured are supported by the resources found : %s are not supported for namespace %s ",
+				strings.Join(unsupportedMetricNames, ","), metric.Namespace)
+		}
+	}
+	if len(supportedMetricNames) == 0 {
+		return nil, errors.Errorf("the metric names configured : %s are not supported for namespace %s ", strings.Join(metric.Name, ","), metric.Namespace)
+	}
+	//validate aggregations and filter on supported ones
+	var supportedAggregations []string
+	var unsupportedAggregations []string
+	metricGroups := make(map[string][]insights.MetricDefinition)
+	metricDefs := getMetricDefinitionsByNames(*metricDefinitions.Value, supportedMetricNames)
+
+	if len(metric.Aggregations) == 0 {
+		for _, metricDef := range metricDefs {
+			metricGroups[string(metricDef.PrimaryAggregationType)] = append(metricGroups[string(metricDef.PrimaryAggregationType)], metricDef)
+		}
+	} else {
+		supportedAggregations, unsupportedAggregations = filterAggregations(metric.Aggregations, metricDefs)
+		if len(unsupportedAggregations) > 0 {
+			return nil, errors.Errorf("the aggregations configured : %s are not supported for some of the metrics selected %s ",
+				strings.Join(unsupportedAggregations, ","), strings.Join(supportedMetricNames, ","))
+		}
+		if len(supportedAggregations) == 0 {
+			return nil, errors.Errorf("no aggregations were found based on the aggregation values configured or supported between the metrics : %s",
+				strings.Join(supportedMetricNames, ","))
+		}
+		metricGroups[strings.Join(supportedAggregations, ",")] = append(metricGroups[strings.Join(supportedMetricNames, ",")], metricDefs...)
```

Comment:
Maybe I am missing something here, but shouldn't the key used be the same?
```suggestion
		key := strings.Join(supportedAggregations, ",")
		metricGroups[key] = append(metricGroups[key], metricDefs...)
```